### PR TITLE
Make EnableTransactionalSession a concern of the persister

### PR DESCRIPTION
**Pull request by danielmarbach**

### CODE REVIEW

File: src/NServiceBus.TransactionalSession.AcceptanceTests/NServiceBus.TransactionalSession.AcceptanceTests.csproj [hunk 1]
```diff
@@ -10,9 +10,9 @@
   <ItemGroup>
     <PackageReference Include="GitHubActionsTestLogger" Version="2.0.1" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.1" />
-    <PackageReference Include="NServiceBus.AcceptanceTesting" Version="8.0.0-rc.1" />
     <PackageReference Include="NUnit" Version="3.13.3" />
     <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
+    <PackageReference Include="NServiceBus.AcceptanceTests.Sources" Version="8.0.0-rc.1" GeneratePathProperty="true" />
```
Comment (timbussmann) [inline src/NServiceBus.TransactionalSession.AcceptanceTests/NServiceBus.TransactionalSession.AcceptanceTests.csproj:15]: we wouldn't have to use all the "remove" parts further down if we don't rely on the sources packages. I think that would overall be simpler at the cost of more explicitly code the default servers to the specific custom persistence

File: src/NServiceBus.TransactionalSession/ConsumeMessageException.cs [hunk 2]
```diff
@@ -2,6 +2,7 @@
 {
     using System;
 
+    [Serializable]
```
Comment (timbussmann) [inline src/NServiceBus.TransactionalSession/ConsumeMessageException.cs:5]: why is that needed?

**Comment by danielmarbach** [comment]
I also tried to make the feature internal but then I couldn't immediately make the ATTs work.

File: src/NServiceBus.TransactionalSession/TransactionalSession.cs [hunk 3]
```diff
@@ -1,33 +1,79 @@
...
+        /// </summary>
+        protected override void Setup(FeatureConfigurationContext context)
         {
-        }
+            QueueAddress localQueueAddress = context.LocalQueueAddress();
 
-        protected override async Task CommitInternal(CancellationToken cancellationToken = default)
-        {
-            await synchronizedStorageSession.CompleteAsync(cancellationToken).ConfigureAwait(false);
+            var isOutboxEnabled = context.Settings.IsFeatureActive(typeof(Outbox));
```
Comment (timbussmann) [inline src/NServiceBus.TransactionalSession/TransactionalSession.cs:23]: would this require an `DependsOnOptionally` (or however this is called) to ensure ordering?

File: src/NServiceBus.TransactionalSession/TransactionalSession.cs [hunk 4]
```diff
@@ -1,33 +1,79 @@
...
+                ITransactionalSession transactionalSession;
+
+                if (isOutboxEnabled)
+                {
+                    transactionalSession = new OutboxTransactionalSession(
+                        sp.GetRequiredService<IOutboxStorage>(),
+                        sp.GetRequiredService<ICompletableSynchronizedStorageSession>(),
+                        sessionCaptureTask.CapturedSession,
+                        sp.GetRequiredService<IMessageDispatcher>(),
+                        sp.GetServices<IOpenSessionOptionsCustomization>(), physicalLocalQueueAddress);
```
Comment (timbussmann) [inline src/NServiceBus.TransactionalSession/TransactionalSession.cs:40]:
```suggestion
                        sp.GetServices<IOpenSessionOptionsCustomization>(),
                        physicalLocalQueueAddress);
```

File: src/NServiceBus.TransactionalSession/ConsumeMessageException.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.TransactionalSession/ConsumeMessageException.cs:5]: technically not needed. I fell into the trap of starting to go down the path of the exception best practices. Happy to remove again

File: src/NServiceBus.TransactionalSession.AcceptanceTests/NServiceBus.TransactionalSession.AcceptanceTests.csproj (hunk 1 above)
Comment (danielmarbach) [reply src/NServiceBus.TransactionalSession.AcceptanceTests/NServiceBus.TransactionalSession.AcceptanceTests.csproj:15]: We use this approach in many persistences as well. At this stage, this is almost an established practice

File: src/NServiceBus.TransactionalSession/ConsumeMessageException.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.TransactionalSession/ConsumeMessageException.cs:5]: Removed

File: src/NServiceBus.TransactionalSession/TransactionalSession.cs (hunk 4 above)
Comment (danielmarbach) [reply src/NServiceBus.TransactionalSession/TransactionalSession.cs:40]: Done

File: src/NServiceBus.TransactionalSession/TransactionalSession.cs (hunk 3 above)
Comment (danielmarbach) [reply src/NServiceBus.TransactionalSession/TransactionalSession.cs]: Done

File: src/NServiceBus.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs [hunk 5]
```diff
@@ -15,7 +16,10 @@ public class TransactionSessionDefaultServer : DefaultServer
             Func<EndpointConfiguration, Task> configurationBuilderCustomization) =>
             base.GetConfiguration(runDescriptor, endpointConfiguration, async configuration =>
             {
-                configuration.EnableTransactionalSession();
+                // Explicitly enforcing the type to be scanned. Otherwise the scanner would not pick it up in the acceptance tests
```
Comment (danielmarbach) [inline src/NServiceBus.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs:19]: @timbussmann I could also make the feature internal but my knowledge with the feature discovery is a bit rusty. I had to force here both things into place; otherwise they won't be taken up by the scanning of the ATTs. That also makes me suspicious of what would happen if we would make the type internal once we try to run ATTs in the downstream persistence repositories. Maybe you have some ideas here?

**Review by tmasternak — APPROVED** [review]

**Review by lailabougria — APPROVED** [review]
The only nitpick I have, is that I still prefer prefixing the testing-specific implementation with AcceptenceTesting.* for clarity, but that's just a preference.

File: src/NServiceBus.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs (hunk 5 above)
Comment (timbussmann) [reply src/NServiceBus.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs:19]: to my understanding `TransactionalSession` shouldn't require to be included explicitly, the default behavior is to exclude all types from the acceptance test repo from scanning (except the current test file) so other features shouldn't be affected.
